keps/sig-apps/2214-array-job/README.md

```diff
@@ -0,0 +1,572 @@
+
+The JobSpec gets the field `CompletionMode` to control whether the Job should
+be treated as an array Job.
+
+```golang
+type JobSpec struct {
+  ...
+
+  // CompletionMode specifies how Pod completions are tracked. It can be
+  // `NonIndexed` (default) or `Indexed`.
+  // `NonIndexed` means that each Pod completion is homologous to each other.
+  // The Job is considered complete when there have been .spec.completions
+  // successful completions.
+  // `Indexed` means that each Pod completion needs to be tracked individually;
+  // each Pods gets an associated completion index, which is part of the Pod
+  // name and it is also available in the annotation
+  // `alpha.batch.k8s.io/job-completion-index`.
+  // The Job is considered complete when there is one successful Pod for each
+  // index in the range 0 to (.spec.completions - 1).
+  // When value is `Indexed`, .spec.completions must have a non-zero positive
+  // value.
+  CompletionMode string
+}
+```
+
+As the comment describes, the `.spec.completions` field needs to be specified
+when `.spec.completionMode = "Indexed"` is used. This is to trigger Job
+management strategy for *fixed completion count*. That is, `Indexed` mode cannot
+be used for work queue patterns.
+
+In the remaining of this KEP, we refer to Jobs with
+`.spec.completionMode = "Indexed"` as Array Jobs.
+
+### Pod details
+
+The Pod and PodSpec APIs don't get any new fields. However, Pods created for
+Array Jobs have two characteristics:
+
+- They get the name `${job-name}-${completion-index}`.
+- They get the annotation `alpha.batch.k8s.io/job-completion-index` with a
+  value equal to its completion index.
+
+Then, the annotation can be accessed through the downward API as a file or
+environment variable.
+
+### Job completion and restart policy
+
+When dealing with Array Jobs, the Job controller keeps track of Pod
+completions for each index from 0 to (`.spec.completions`-1). When a Pod finishes
+successfully, no other Pod is created with the same completion index.
+Succeeded Pods are not removed by the Job controller until the entire Job gets
+removed.
+
+If an entire Pod fails, such as when the Pod is kicked off the node or if a
+container of the Pod fails and `.spec.template.spec.restartPolicy = "Never"`,
+the Job controller first removes the old Pod and then creates a new Pod with the
+same name. The application needs to handle restarts in a different Pod.
+
+<<[UNRESOLVED Keeping track of number of failures]>>
+In non-array Jobs, the Job controller reports the number of failures in the
+Job status from the number of existing failed Pods.
+
+Since array Jobs remove the failed Pods, the information needs to be tracked in
+other ways.
+One solution is keeping a recreation counter as an annotation in the Pod:
+`alpha.batch.k8s.io/job-pod-recreate-count`. Then, before recreating a Pod,
+the counter is extracted, incremented and added to the new Pod. The counters
+are the source-of-truth for Job failures.
```

Comment:
What if the conroller gets restarted between removing the previous pod and creating the new one?

Reply:
That would be an scenario where information is lost.

However, note that we also loose information in the current architecture: when the Pod garbage collector reaches the threshold and starts removing failed (or even succeeded) Pods.

Reply:
We decided to remove stable names from the proposal. It might be reconsidered as a future related enhancement, after kubernetes/kubernetes#28486 is resolved and from user feedback.

I moved the details to "Alternatives"